pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1547,6 +1561,93 @@ func useCSIPlugin(vpm *volume.VolumePluginMgr, spec *volume.Spec) bool {
 	return false
 }
 
+func nodeUsingCSIPlugin(og *operationGenerator, spec *volume.Spec, nodeName types.NodeName) (bool, error) {
+	var err error
+
+	migratable, err := og.volumePluginMgr.IsPluginMigratableBySpec(spec)
+	if err != nil {
+		return false, err
+	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!migratable {
+		return false, nil
+	}
+
+	if len(nodeName) == 0 {
+		return false, fmt.Errorf("nodeName is empty")
+	}
+
+	vpm := og.volumePluginMgr
+
+	kubeClient := vpm.Host.GetKubeClient()
+	if kubeClient == nil {
+		// TODO(dyzz) check this error case, what should we do in standalone kubelet mode?
+		return false, fmt.Errorf("failed to get kube client from volume host")
+	}
+
+	adcHost, ok := og.volumePluginMgr.Host.(volume.AttachDetachVolumeHost)
+	if !ok {
+		// This function is running not on the AttachDetachController
+		// We assume that Kubelet is servicing this function and therefore is
+		// trivially "using CSI Plugin"
+		return true, nil
+	}
+	var csiNode *storagev1beta1.CSINode
+	if adcHost.CSINodeSynced() {
```

Comment:
Should we be waiting for the informer to be synced in ADController.Run() instead? (that's where we wait for everything else to be synced)

Reply:
that is probably wise.